src/Database/Connection.php

```diff
@@ -178,29 +181,48 @@ public function configName(): string
      * Sets the driver instance. If a string is passed it will be treated
      * as a class name and will be instantiated.
      *
-     * @param \Cake\Database\DriverInterface|string $driver The driver instance to use.
+     * @param \Cake\Database\Driver|string $driver The driver instance to use.
      * @param array<string, mixed> $config Config for a new driver.
      * @throws \Cake\Database\Exception\MissingDriverException When a driver class is missing.
      * @throws \Cake\Database\Exception\MissingExtensionException When a driver's PHP extension is missing.
      * @return $this
+     * @deprecated 4.4.0 Setting the driver is deprecated. Use the connection config instead.
      */
     public function setDriver($driver, $config = [])
     {
+        deprecationWarning('Setting the driver is deprecated. Use the connection config instead.');
+
+        $this->_driver = $this->createDriver($driver, $config);
+
+        return $this;
+    }
+
+    /**
+     * Creates driver from name, class name or instance.
+     *
+     * @param \Cake\Database\Driver|string $name Driver name, class name or instance.
+     * @param array $config Driver config if $name is not an instance.
+     * @return \Cake\Database\Driver
+     * @throws \Cake\Database\Exception\MissingDriverException When a driver class is missing.
+     * @throws \Cake\Database\Exception\MissingExtensionException When a driver's PHP extension is missing.
+     */
+    protected function createDriver($name, array $config): Driver
+    {
+        $driver = $name;
         if (is_string($driver)) {
-            /** @psalm-var class-string<\Cake\Database\DriverInterface>|null $className */
+            /** @psalm-var class-string<\Cake\Database\Driver>|null $className */
```

Comment:
Why did require to be changed?

Reply:
`createDriver()` returns a `Driver` instance instead of `DriverInterface`. Perhaps it's too early to make those declarations, but it's already a requirement.

Reply:
So why not use the interface for return type since `Driver` implements `DriverInterface`?

Reply:
It's a step towards acknowledging we require drivers to extend Driver. I can change it for this PR though.

Reply:
Ah, better to do that in PR which actually removes the use of DriverInterface.